### ISSUE TITLE
FSE submission TODOs

### ISSUE THREAD

**Issue by shuaiwang516** [xlab-uiuc/confuzz#182]
# Understanding Coverage
- [ ] Figure out why in some tests/projects, the start point of coverage of Confuzz is much lower than JQF and Random
    - [X] @shuaiwang516  - Some tests fail under the default configuration value (either due to test bug or flakiness). In this case, Confuzz always fuzz default, but JQF and Random try other things and increase coverage (trigger those checking failures so that the coverage is higher than Confuzz) - this can be potentially solved by running the pre-round twice, and we only fuzz those tests that pass both default round. Done in https://github.com/xlab-uiuc/confuzz/commit/107f11467cbebe31da6a12d55e4f4c47f436a044
    - [ ] @shuaiwang516  - Some test pass - do not know the reason here, need further inspection
    - [ ] **`Medium-Priority`** Try to figure out the Union Coverage calculation 
- [ ] Figure out why some tests can not produce coverage reports correctly
    - [ ] @hwang-pku - This may be due to the Azure VM resource limitation, because we sampled some cases to inspect, and the coverage calculation only failed on Azure VM but succeeded on our server and local machine. How to solve? We can either rerun the coverage goal on local, or launch another coverage round on Azure with more powerful machines that have more RAM (But I guess this would be very expensive). Do we have another way?
    - [ ] What are the other reasons?
- [ ] Figure out what branches are covered by Confuzz, but not by JQF and Random; and vice versa.

# Understanding Bugs and Failures
- [ ] @sampalomad @shuaiwang516 @hwang-pku Labeling failures to bugs that Confuzz have found
- [ ] Figure out what bugs can be found by Confuzz, but missed in JQF and Random. 

# Understanding Crashes of Fuzzing
- [ ] Why do some tests never fuzz any inputs? or only one?
     - [ ] test timeout after 15 min?

Please feel free to edit / add more TODOs.

**Comment by tianyin** [comment]
@shuaiwang516 I think the assumption should be that the correct tests should always pass with the default configuration. That's also the requirement made by Ctest. If a test fails on the default configuration, it basically means the test is incorrect (mostly likely that the environment is wrong).

**Comment by shuaiwang516** [comment]
> @shuaiwang516 I think the assumption should be that the correct tests should always pass with the default configuration. That's also the requirement made by Ctest. If a test fails on the default configuration, it basically means the test is incorrect (mostly likely that the environment is wrong).

Exactly! I totally agree that we should only fuzz the tests that can pass under the default configuration.
I thought I did this before, but apparently, my code doesn't work like that ;(

I modified the logic of our pre-round, so right now Confuzz checks the test execution under the default configuration **twice** before fuzzing. We only continue fuzzing if both execution pass. Otherwise either the test code is incorrect, or the test itself is flaky.

**Comment by darko-marinov** [comment]
The (flaky) tests that fail when run in isolation (by passing with other tests) were called `brittle` (by al. et Marinov). We found a number of such tests in open-source projects. Developers sometimes accept fixes so that tests can run in isolation.

For this project, it's fine to say that some tests were flaky, and we ignored them.

**Comment by tianyin** [comment]
> https://github.com/xlab-uiuc/confuzz/issues/182#issuecomment-1664823778

@shuaiwang516 so did your coverage increase now?

**Comment by shuaiwang516** [comment]
> > [#182 (comment)](https://github.com/xlab-uiuc/confuzz/issues/182#issuecomment-1664823778)
> 
> @shuaiwang516 so did your coverage increase now?

For those tests that fail at pre-round, if we remove it, the coverage definitely would increase. 
Though we didn't draw the figure, apparently this is true because we essentially remove some tests that Confuzz cov is lower than JQF/Random.

On the other hand, I'm debugging why some tests fail. I found that in the vanilla surefire round test can pass, but with our tool, it fails. The initial guess is that there are some project/test-specific environment arguments or system properties the tool does not handle well.
